Automattic/woocommerce-payments
# RPP - Connect the duplicate payment prevention service (Verification)

Fixes #7411

## Changes proposed in this Pull Request

- Create new `DuplicatePaymentPreventionService` class in `src` and keep most of logics from the legacy one.
- Hooks: remove the hook from old legacy service, and register hooks from the new `src` service. 
- Introduce only one new state `DuplicateOrderDetectedState` when a duplicate order is detected.
- Introduce `Intent_Status::AUTHORIZED_STATUSES` and `is_authorized()` for intent abstraction.
- Add `SessionService` in `src` with the purpose of easily handling session data (sometimes, it's null while other times, it's really the expected instance of `WC_Session`). This will also be useful when other features will use it, such as Fraud relevant features.
- Add tests, cover 100% for `src` files.









## Testing instructions 

- At this stage, we can only test for: 
    - type of card: basic. 3DS is not yet supported: though #7471 is merged but it's still not possible to finish checkout with a 3DS card - see more details p1698122299758599-slack-C05997J5PCG
    - checkout page type: both classic and block.
    - method: both new and saved method.
    - gateway type: does not matter as we only have one type for the new payment process. 

- Ensure that you're using the latest versions of Dev Tools and enable **New Payment Process Factors**.
- Ideally, add Xdebug breakpoints in the new code in InitialState.php and the new service php file to confirm new code really executed. 

### For testing duplicate orders 

Basically, follow test instruction and result of this PR https://github.com/Automattic/woocommerce-payments/pull/5247 with a note: 
- Use this updated snippet https://gist.github.com/htdat/757b1375146584e63ebad80c5d02f1a1 (rather than the one in PR 5247) 

### For testing duplicate payments 

Basically, follow test instruction and result of this PR https://github.com/Automattic/woocommerce-payments/pull/5346 with a note: 

- For the diff, please use this new one https://gist.github.com/htdat/14715ca445cea3d4d704d5d1a924b25f
- For the snippet, it's OK to use the one mentioned in PR 5346 https://gist.github.com/htdat/6a52eec1be77b8442ac05401908a3627






## Remaining TODO(s) for this PR: 

N/A

## Finished TODOs for this PR

- [x] Consider port this change https://github.com/Automattic/woocommerce-payments/pull/7441 - We may not need this as the fix for PR 7441 is for legacy UPE, and split UPE (not dUPE), which are going to be deprecated. 
- [x] Use `Logger` as dependency per https://github.com/Automattic/woocommerce-payments/issues/7428
- [x] Consider including this refactoring https://github.com/Automattic/woocommerce-payments/issues/6712
- [x] Register hooks but avoid duplicate hooks https://github.com/Automattic/woocommerce-payments/pull/7450#discussion_r1356675850
- [x] Update to use a post-procesing state https://github.com/Automattic/woocommerce-payments/pull/7450#discussion_r1363170224

-------------------


**Post merge**
